### ISSUE TITLE
regression: dbuild is picky about the format of the `version` settings in projects it builds

### ISSUE THREAD

**Issue by SethTisue** [typesafehub/dbuild#182]
we recently upgraded 2 of the 3 Scala community builds to from dbuild 0.9.1 to 0.9.4.

in the 2.12 community build, we are now seeing this failure (excerpted from [this log](https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-integrate-community-build/241/consoleFull)):

```
[scalaz-stream] --== Building scalaz-stream ==--
[scalaz-stream] Resolving: https://github.com/scalaz/scalaz-stream.git#60eecfdd2e29ba2a789ae411a8653e6201ca3e17 in directory: ./target-0.9.4/project-builds/scalaz-stream-765eea1747f324b13dc0ed707590ff048337a07e
[scalaz-stream] Fetching https://github.com/scalaz/scalaz-stream.git
[scalaz-stream] into /home/jenkins/workspace/scala-2.12.x-integrate-community-build/target-0.9.4/clones/67ec136f8a1eb6a9e0634a1213c79ea659ab521b
[scalaz-stream] Took: 00h 00m 00.7s
[scalaz-stream] Cloning /home/jenkins/workspace/scala-2.12.x-integrate-community-build/target-0.9.4/clones/67ec136f8a1eb6a9e0634a1213c79ea659ab521b
[scalaz-stream] to /home/jenkins/workspace/scala-2.12.x-integrate-community-build/target-0.9.4/project-builds/scalaz-stream-765eea1747f324b13dc0ed707590ff048337a07e
[scalaz-stream] Took: 00h 00m 00.0s
[scalaz-stream] Fetching /home/jenkins/workspace/scala-2.12.x-integrate-community-build/target-0.9.4/clones/67ec136f8a1eb6a9e0634a1213c79ea659ab521b
[scalaz-stream] into /home/jenkins/workspace/scala-2.12.x-integrate-community-build/target-0.9.4/project-builds/scalaz-stream-765eea1747f324b13dc0ed707590ff048337a07e
[scalaz-stream] Took: 00h 00m 00.1s
[scalaz-stream] Resolving artifacts, level 0, space: default
[scalaz-stream] Retrieved from project scala (commit: none): 12 artifacts
[scalaz-stream] Retrieved from project scalacheck (commit: 2809d8d5616eebaa365c890717d30fd8a28368da): 4 artifacts
[scalaz-stream] Retrieved from project scalatest (commit: bff5bd8e9ee874699189e2a1f851716737d91306): 4 artifacts
[scalaz-stream] Retrieved from project scalaz (commit: 727761730750ac611fcb9d91d6491ae235874c09): 29 artifacts
[scalaz-stream] Retrieved from project scodec-bits (commit: bceb361e327a595fbd0a81bf55901195882f382d): 4 artifacts
[scalaz-stream:error] scala.MatchError: snapshot-0.7a (of class java.lang.String)
[scalaz-stream:error] 	at com.typesafe.dbuild.project.build.LocalBuildRunner$.prepareDepsArtifacts(LocalBuildRunner.scala:109)
[scalaz-stream:error] 	at com.typesafe.dbuild.project.build.LocalBuildRunner$$anonfun$runLocalBuild$1.apply(LocalBuildRunner.scala:56)
[scalaz-stream:error] 	at com.typesafe.dbuild.project.build.LocalBuildRunner$$anonfun$runLocalBuild$1.apply(LocalBuildRunner.scala:46)
[scalaz-stream:error] 	at com.typesafe.dbuild.project.build.BuildDirs$.useProjectUniqueBuildDir(BuildDirs.scala:39)
[scalaz-stream:error] 	at com.typesafe.dbuild.project.build.LocalBuildRunner.runLocalBuild(LocalBuildRunner.scala:46)
[scalaz-stream:error] 	at com.typesafe.dbuild.project.build.LocalBuildRunner.checkCacheThenBuild(LocalBuildRunner.scala:35)
[scalaz-stream:error] 	at com.typesafe.dbuild.project.build.BuildRunnerActor$$anonfun$receive$2$$anonfun$apply$1.apply$mcV$sp(BuildRunnerActor.scala:40)
[scalaz-stream:error] 	at com.typesafe.dbuild.project.build.BuildRunnerActor$$anonfun$receive$2$$anonfun$apply$1.apply(BuildRunnerActor.scala:38)
[scalaz-stream:error] 	at com.typesafe.dbuild.project.build.BuildRunnerActor$$anonfun$receive$2$$anonfun$apply$1.apply(BuildRunnerActor.scala:38)
[scalaz-stream:error] 	at com.typesafe.dbuild.project.controller.Controller$.forwardingErrorsToFuturesControlled(Controller.scala:59)
[scalaz-stream:error] 	at com.typesafe.dbuild.project.build.BuildRunnerActor$$anonfun$receive$2.apply(BuildRunnerActor.scala:38)
[scalaz-stream] --== End Building scalaz-stream ==--
```

@cunei, what do you make of this? perhaps the regex in 0a7f12a7bb505965e3042bb7f88d885724537bb3 isn't quite right?

in the project in question (scalaz-stream), build.sbt has `version := "snapshot-0.7a"`

**Comment by cunei** [comment]
If I had to make a guess, "0.7a" is not a proper Maven version number. It should be "0.7-a", or "0.7.1", but "0.7a" breaks ordering when Maven checks which version is newer. Some details here: https://cwiki.apache.org/confluence/display/MAVENOLD/Dependency+Mediation+and+Conflict+Resolution

**Comment by SethTisue** [comment]
It's not some maven thing being picky here; it's dbuild itself.

I think it's the `snapshot-` at the start, actually; the regex dbuild is using assumes there's a number at the start.

I would prefer that dbuild have graceful fallback behavior in this case, since in general in the community build we have to cope with all sorts of people doing odd things in their builds. But anyway, we only have one project it's breaking on, so I will work around it by forking that project.

(I tried working around it by doing `set version ...` in `extra.commands` but the error happens too early for that to help.)

**Comment by cunei** [comment]
@SethTisue, I worked on a patch (bc11e2f) that would allow compilation to proceed in case the version string is incorrectly formatted, as in this case. However, I am not convinced that merging it would be a good idea: considering that much of what dbuild does revolves around version number rewriting, working with version strings that do not conform to the format that many tools expect may lead to all sort of problems. I am rather inclined to keep the current behavior, instead, especially since the version string can always be fixed in a separate fork.

**Comment by SethTisue** [comment]
OK. I'll open a ticket with the scalaz-stream folks. perhaps this won't happen often.

**Comment by SethTisue** [comment]
ha. they've actually already fixed it in both their 0.8 and 0.9 series branches.

**Comment by SethTisue** [comment]
closing, but perhaps you might improve the error message here?
